lsp-mode.el

```diff
@@ -5515,40 +5515,34 @@ A reference is highlighted only if it is visible in a window."
                                       line
                                       character)))
 
-(defun lsp--get-document-symbols ()
+(defun lsp--get-document-symbols (&optional callback)
   "Get document symbols.
 
 If the buffer has not been modified since symbols were last
 retrieved, simply return the latest result.
 
-Else, if the request was initiated by Imenu updating its menu-bar
-entry, perform it asynchronously; i.e., give Imenu the latest
-result and then force a refresh when a new one is available.
+If CALLBACK is specified, call it when the document's symbols
+have been fetched. It implies
+`lsp--document-symbols-request-async'.
 
-Else (e.g., due to interactive use of `imenu' or `xref'),
-perform the request synchronously."
+Else (e.g., due to interactive use of `imenu' or `xref'), perform
+the request synchronously."
   (if (= (buffer-chars-modified-tick) lsp--document-symbols-tick)
       lsp--document-symbols
     (let ((method "textDocument/documentSymbol")
           (params `(:textDocument ,(lsp--text-document-identifier)))
           (tick (buffer-chars-modified-tick)))
-      (if (not lsp--document-symbols-request-async)
-          (prog1
-              (setq lsp--document-symbols (lsp-request method params))
-            (setq lsp--document-symbols-tick tick))
-        (lsp-request-async method params
-                           (lambda (document-symbols)
-                             (setq lsp--document-symbols document-symbols
-                                   lsp--document-symbols-tick tick)
-                             (lsp--imenu-refresh))
-                           :mode 'alive)
+      (if (or callback lsp--document-symbols-request-async)
+          (lsp-request-async method params
+                             (lambda (document-symbols)
+                               (setq lsp--document-symbols document-symbols
+                                     lsp--document-symbols-tick tick)
+                               (when callback (funcall callback)))
+                             :mode 'alive)
+        (setq lsp--document-symbols (lsp-request method params)
+              lsp--document-symbols-tick tick)
         lsp--document-symbols))))
 
-(advice-add 'imenu-update-menubar :around
```

Comment:
Please avoid doing changes like that - you cannot go and just delete a code like that. This will cause the imenu to be loaded synchronously and emacs will block until the server has started which may be minutes.

Reply:
This won't cause issues. Note that there now is a callback parameter, which is specified for `lsp--imenu-create-index`. It implies asynchronous operation. The callback is `lsp--imenu-refresh` for `lsp--imenu-create-index`.

Reply:
No - you are introducing a regression. lsp--imenu-create-index function is called in 2 contexts - one is async and one is sync.  It is not calling lsp--imenu-refresh only if it is called from the menubar, so from the callers perspective there is no difference.

Reply:
> . This will cause the imenu to be loaded synchronously and emacs will block until the server has started which may be minutes.

(The issue is actually you will break the sync path, not the async one).